controller/handler/badge.go

```diff
@@ -35,22 +35,17 @@ var (
 
 // UptimeBadge handles the automatic generation of badge based on the group name and endpoint name passed.
 //
-// Valid values for {duration}: 7d, 24h, 1h
+// Duration should be in the form of a number with date marker, e.g. 6h,2d,3m,1y
 func UptimeBadge(writer http.ResponseWriter, request *http.Request) {
 	variables := mux.Vars(request)
 	duration := variables["duration"]
-	var from time.Time
-	switch duration {
-	case "7d":
-		from = time.Now().Add(-7 * 24 * time.Hour)
-	case "24h":
-		from = time.Now().Add(-24 * time.Hour)
-	case "1h":
-		from = time.Now().Add(-2 * time.Hour) // Because uptime metrics are stored by hour, we have to cheat a little
-	default:
-		http.Error(writer, "Durations supported: 7d, 24h, 1h", http.StatusBadRequest)
+
+	from, err := parseFromTime(duration)
+	if err != nil {
+		http.Error(writer, err.Error(), http.StatusBadRequest)
```

Comment:
The reason why I allowed only specific durations to be passed as parameter is because calculating the uptime can be expensive, and I'd like that to stay that way.

I'm fine with adding the following values:
- 30d
- 90d
- 180d
- 1y

but we'll have to add a caching mechanism.

Reply:
You are right, unrestricted queries can lead to dos attacks

Reply:
I've come up with these strategies

1. adjust `uptimeCleanUpThreshold` to 1 year, the disadvantage is that it will take up storage space and the query will be slower
2. allow the user to specify the allowed `durations` through the config file, `uptimeCleanUpThreshold` is determined by the maximum `duration`.
3. allow the user to specify the allowed `durations` and `uptimeCleanUpThreshold` through the config file, and  user should ensure that `uptimeCleanUpThreshold` is not lower than the maximum `duration`

I think the second strategy is better, do you have any suggestions?